src/vs/workbench/contrib/terminal/browser/xterm/markNavigationAddon.ts

```diff
@@ -40,62 +38,63 @@ export class CommandNavigationAddon extends Disposable implements ICommandTracke
 	}
 
 	constructor(
-		store: ITerminalCapabilityStore,
+		private readonly _capabilityStore: ITerminalCapabilityStore,
 		@IThemeService private readonly _themeService: IThemeService
 	) {
 		super();
-		this._refreshActiveCapability(store);
-		this._register(store.onDidAddCapability(() => this._refreshActiveCapability(store)));
-		this._register(store.onDidRemoveCapability(() => this._refreshActiveCapability(store)));
 	}
 
-	private _refreshActiveCapability(store: ITerminalCapabilityStore) {
-		const activeCommandDetection = store.get(TerminalCapability.CommandDetection) || store.get(TerminalCapability.PartialCommandDetection);
-		if (activeCommandDetection !== this._commandDetection) {
-			this._commandDetection = activeCommandDetection;
-		}
-	}
-
-	private _getCommandMarkers(skipEmptyCommands: boolean = false): readonly IMarker[] {
-		if (!this._commandDetection) {
-			return [];
-		}
-		let commands: readonly IMarker[];
-		if (this._commandDetection.type === TerminalCapability.PartialCommandDetection) {
-			commands = this._commandDetection.commands;
-		} else {
-			commands = coalesce(this._commandDetection.commands.filter(e => skipEmptyCommands ? e.command : true).map(e => e.marker));
+	private _getMarkers(skipEmptyCommands?: boolean): readonly IMarker[] {
+		const commandCapability = this._capabilityStore.get(TerminalCapability.CommandDetection);
+		const partialCommandCapability = this._capabilityStore.get(TerminalCapability.PartialCommandDetection);
+		const markCapability = this._capabilityStore.get(TerminalCapability.BufferMarkDetection);
+		let markers: IMarker[] = [];
+		if (commandCapability) {
+			markers = coalesce(commandCapability.commands.map(e => e.marker));
+		} else if (partialCommandCapability) {
+			markers.push(...partialCommandCapability.commands);
+		}
+
+		if (markCapability && !skipEmptyCommands) {
+			let next = markCapability.markers().next()?.value;
+			const arr: IMarker[] = [];
+			while (next) {
+				arr.push(next);
+				next = markCapability.markers().next()?.value;
+			}
+			markers = arr;
 		}
-		return commands;
+		return markers;
 	}
 
 	clearMarker(): void {
 		// Clear the current marker so successive focus/selection actions are performed from the
 		// bottom of the buffer
 		this._currentMarker = Boundary.Bottom;
-		this._resetNavigationDecoration();
+		this._resetNavigationDecorations();
 		this._selectionStart = null;
 	}
 
-	private _resetNavigationDecoration() {
-		this._navigationDecoration?.dispose();
-		this._navigationDecoration = undefined;
+	private _resetNavigationDecorations() {
+		if (this._navigationDecorations) {
+			dispose(this._navigationDecorations);
+		}
```

Comment:
`dispose` takes undefined so this should work:

```suggestion
		dispose(this._navigationDecorations);
```